src/lib/symbol_overrides.cc

```diff
@@ -67,22 +80,168 @@ void check_libraries() {
   }
 }
 
+struct AllocTrackerHelper {
+  AllocTrackerHelper()
+      : tl_state{ddprof::AllocationTracker::get_tl_state()},
+        guard{tl_state ? &(tl_state->reentry_guard) : nullptr} {
+    check_libraries();
+  }
+
+  void track(void *ptr, size_t size) {
+    if (guard) {
+      ddprof::AllocationTracker::track_allocation_s(
+          reinterpret_cast<uintptr_t>(ptr), size, *tl_state);
+    }
+  }
+
+  ddprof::TrackerThreadLocalState *tl_state;
+  ddprof::ReentryGuard guard;
+};
+
+struct DeallocTrackerHelper {
+  DeallocTrackerHelper()
+      : tl_state{ddprof::AllocationTracker::is_deallocation_tracking_active()
+                     ? ddprof::AllocationTracker::get_tl_state()
+                     : nullptr},
+        guard{tl_state ? &(tl_state->reentry_guard) : nullptr} {}
+
+  void track(void *ptr) {
+    if (guard) {
+      ddprof::AllocationTracker::track_deallocation_s(
+          reinterpret_cast<uintptr_t>(ptr), *tl_state);
+    }
+  }
+
+  ddprof::TrackerThreadLocalState *tl_state;
+  ddprof::ReentryGuard guard;
+};
+
 struct malloc {
   static constexpr auto name = "malloc";
   static inline auto ref = &::malloc;
   static inline bool ref_checked = false;
 
   static void *hook(size_t size) noexcept {
-    check_libraries();
-    ddprof::TrackerThreadLocalState *tl_state =
-        ddprof::AllocationTracker::get_tl_state();
-    ddprof::ReentryGuard guard(tl_state ? &(tl_state->double_tracking_guard)
-                                        : nullptr);
+    AllocTrackerHelper helper;
```

Comment:
so the assumption is that we always keep track of the first function that is tracked (as that is the one that keeps the guard)
This is good for 
```
malloc -> mmap
free -> munmap
```
Are we worried about things that would happen in a reverse order ?

Reply:
Are you worried about `mmap` calling `malloc` ? I don't see how that could happen (mmap is a wrapper around a syscall).